tests/spiders.py

```diff
@@ -28,6 +28,45 @@ def closed(self, reason):
         self.meta['close_reason'] = reason
 
 
+class KeywordArgumentsSpider(MockServerSpider):
```

Comment:
Could you please add a test which checks how Scrapy behaves if there is a mismatch between parameters `parse` accept and parameters passed via callback kwargs, e.g. some required argument is missing, or an extra argument is passed? Another case which could be nice to check explicitly is how default values are handled.

Reply:
Added tests for defaults and argument mismatch.
Errback is not called with this current implementation, it does if we add the callback and the errback to the Deferred in two separate steps, i.e.
```python
dfd.addCallback(request.callback or spider.parse, **request.cb_kwargs)
dfd.addErrback(request.errback)
```
instead of
```python
dfd.addCallbacks(
    callback=request.callback or spider.parse,
    errback=request.errback,
    callbackKeywords=request.cb_kwargs)
```
In any case, the `Request` object is not bounded to the `Failure` received by the errback. Personally, I don't think calling the errback would be appropriate here, since it's not an error with the request/response itself, but with the code that handles it. The logged error is very descriptive, and similar to the one that currently appears when, for instance, a callback does not take a second positional argument (`TypeError: parse() takes 1 positional argument but 2 were given`).